### PR TITLE
Expose some GIL functions

### PR DESCRIPTION
**Pull request by LaurentMazare**
This PR exposes a couple functions from the Python C api related to the global interpreter lock. Using them makes it possible to run some python functions from different threads (in our use case we have a python process that runs the async scheduler in a separate pthread so that it can run tasks in the background while some python code is running).

### CODE REVIEW

**Comment by LaurentMazare** [comment]
@thierry-martinez any thoughts on this ?

**Comment by LaurentMazare** [comment]
Ping @thierry-martinez ?

File: py.ml [hunk 1]
```diff
@@ -2414,6 +2414,17 @@ module Gil = struct
   let ensure = Pywrappers.pygilstate_ensure
   let release = Pywrappers.pygilstate_release
   let check () = Pywrappers.pygilstate_check () <> 0
+
+  let with_lock f =
```
Comment (thierry-martinez) [inline py.ml]: I think a simpler implementation would be
```ocaml
let t = ensure () in
Fun.protect f ~finalize:(fun () -> release t)
```
Moreover, `Fun.protect` takes care about preserving backtraces.

File: py.mli [hunk 2]
```diff
@@ -1988,6 +1988,23 @@ module Array: sig
       you should use bigarrays and the {!Numpy} module. *)
 end
 
+module Gil : sig
+  type t
+
+  val ensure : unit -> t
+  (** [ensure ()] ensures that the current thread holds the global
+      interpreter lock and hence can call the Python C API in a safe
+      way.
```
Comment (thierry-martinez) [inline py.mli]: I would add a link to https://docs.python.org/3/c-api/init.html#c.PyGILState_Ensure

File: py.mli [hunk 3]
```diff
@@ -1988,6 +1988,23 @@ module Array: sig
...
+  *)
+
+  val release : t -> unit
+  (** [release t] releases any resource acquired by [ensure]. *)
```
Comment (thierry-martinez) [inline py.mli]: https://docs.python.org/3/c-api/init.html#c.PyGILState_Release

File: py.mli [hunk 4]
```diff
@@ -1988,6 +1988,23 @@ module Array: sig
...
+
+  val check : unit -> bool
+  (** [check ()] returns true if the current thread holds the global
+      interpreter lock. *)
```
Comment (thierry-martinez) [inline py.mli]: https://docs.python.org/3/c-api/init.html#c.PyGILState_Check

File: dune [hunk 5]
```diff
@@ -2,9 +2,8 @@
   (name pyml)
   (modules numpy py pycaml pyml_arch pytypes pywrappers)
   (c_names numpy_stubs pyml_stubs)
-  (flags :standard -thread)
   (wrapped false)
-  (libraries bigarray stdcompat pyutils threads.posix))
+  (libraries bigarray stdcompat pyutils))
```
Comment (thierry-martinez) [inline dune]: Good! I would have been reluctant to have `threads.posix` be a dependency of `pyml`. If such a change is required in the future, I would prefer to isolate this need in a `pyml.threads` sub-package.

**Comment by LaurentMazare** [comment]
Thanks for the comments, I just pushed some changes to fix the issues you mentioned.

**Comment by thierry-martinez** [comment]
Thank you very much! Merged.

**Pull request merged**